### PR TITLE
09 - added shopping cart and cart item

### PR DESCRIPTION
**Pull request by yevhenkukharchuk**

### CODE REVIEW

File: src/main/java/yevhen/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    @PreAuthorize("hasRole('USER')")
+    public CartItemDto addBookToShoppingCart(@RequestBody @Valid CartItemRequestDto requestDto) {
+        return shoppingCartService.addCartItem(requestDto);
+    }
+
+    @PutMapping("/cart-items/{cartItemId}")
+    @Operation(summary = "Update quantity of a book",
+            description = "Update quantity of a book in the shopping cart")
+    @PreAuthorize("hasRole('USER')")
+    public CartItemDto updateQuantity(@PathVariable @Positive Long cartItemId,
```
Comment (ystankevych) [inline src/main/java/yevhen/bookstore/controller/ShoppingCartController.java:52]: @Positive will not work without @Validated annotation placed above the class (so just add @Validated annotation on this controller)

File: src/main/java/yevhen/bookstore/model/ShoppingCart.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+@Entity
+@Getter
+@Setter
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @OneToOne(fetch = FetchType.LAZY, optional = false)
```
Comment (ystankevych) [inline src/main/java/yevhen/bookstore/model/ShoppingCart.java:25]: I would suggest using @MapsId to share the user id for the shopping cart (but in that case you need to remove @GeneratedValue annotation from id field(here in this class)). But its up to you =)

**Review by ystankevych — CHANGES_REQUESTED** [review]
Hi! Great job. I just left some comments

File: src/main/java/yevhen/bookstore/model/ShoppingCart.java (hunk 2 above)
Comment (yevhenkukharchuk) [reply src/main/java/yevhen/bookstore/model/ShoppingCart.java:25]: thank you for advice, it is good improvement

File: src/main/java/yevhen/bookstore/mapper/ShoppingCartMapper.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS
+)
+public interface ShoppingCartMapper {
+    @Mapping(target = "userId", source = "user.id")
+    ShoppingCartDto toDto(ShoppingCart shoppingCart);
+
+    @Mapping(target = "bookId", source = "book.id")
+    @Mapping(target = "bookTitle", source = "book.title")
+    CartItemDto toDto(CartItem cartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/yevhen/bookstore/mapper/ShoppingCartMapper.java]: This mapping should be in the CartItemMapper class.

File: src/main/java/yevhen/bookstore/mapper/ShoppingCartMapper.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import org.mapstruct.Mapping;
+import org.mapstruct.NullValueCheckStrategy;
+import yevhen.bookstore.dto.cartitem.CartItemDto;
+import yevhen.bookstore.dto.shoppingcart.ShoppingCartDto;
+import yevhen.bookstore.model.CartItem;
+import yevhen.bookstore.model.ShoppingCart;
+
+@Mapper(componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS
```
Comment (Elena-Bruyako) [inline src/main/java/yevhen/bookstore/mapper/ShoppingCartMapper.java]:
```suggestion
        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
        uses = CartItemMapper.class
```

File: src/main/java/yevhen/bookstore/model/ShoppingCart.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+@Entity
+@Getter
+@Setter
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    private Long id;
+
+    @OneToOne(fetch = FetchType.LAZY, optional = false)
```
Comment (Elena-Bruyako) [inline src/main/java/yevhen/bookstore/model/ShoppingCart.java:25]: `@JoinColumn(name = "id", nullable = false)`

File: src/main/java/yevhen/bookstore/model/ShoppingCart.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    @MapsId
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/yevhen/bookstore/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/resources/db/changelog/changes/09-create-cart-items-table.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                  constraints:
+                    nullable: false
+                    foreignKeyName: foreignKey_shopping_cart_id
+                    references: shopping_carts(id)
+              - column:
+                  name: book_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+                    foreignKeyName: frgnKey_book_id
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/09-create-cart-items-table.yml]: let's be consistent with naming like foreignKey_shopping_cart_id

File: src/main/resources/db/changelog/changes/08-create-shopping-carts-table.yml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    nullable: false
+                    primaryKey: true
+              - column:
+                  name: user_id
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/08-create-shopping-carts-table.yml]: As you are using MapsId annotation you don't need to create foreignKey here

File: src/main/resources/db/changelog/changes/09-create-cart-items-table.yml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                    references: books(id)
+              - column:
+                  name: quantity
+                  type: int
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/09-create-cart-items-table.yml:34]: add empty line 
Check all changelogs

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/yevhen/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Override
+    public void deleteCartItemById(Long id) {
+        ShoppingCart currentCart = getCurrentCurt();
+        cartItemRepository.findByIdAndShoppingCartId(id, currentCart.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Can't get cart item by id:" + id));
+        cartItemRepository.deleteById(id);
+    }
+
+    private ShoppingCart getCurrentCurt() {
```
Comment (liliia-ponomarenko) [inline src/main/java/yevhen/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java:68]:
```suggestion
    private ShoppingCart getCurrentCart() {
```

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😊

**Pull request merged**